### PR TITLE
[Tabs] Apply design system

### PR DESCRIPTION
**Pull request by LauraAubin**
<details>
<summary>To-do list</summary>

- [x] No focus ring when clicked on with mouse (tab index -1 on inner element)
- [x] Clean up legacy css spaghetti (if there is time)
- [x] De-globalize `newDesignSystem`
- [x] Stop using a css var for container padding
- [x] Use :after for the underline
- [x] Test coverage
- [x] Underline (absolute, avoid off by 1)
</details>

### WHY are these changes introduced?

- Part of https://github.com/Shopify/polaris-ux/issues/364
- [Figma](https://www.figma.com/file/4dAAt5iFPSaxUKiYVKrkYj/DL%E2%80%93Polaris-(Web)%3A-Components?node-id=898%3A0)
- [CSS Layout in Codepen](https://codepen.io/LauraAubin/pen/rNVmbJR?editors=1100)

### WHAT is this pull request doing?

Applies the design language to the Tabs component.

<kbd>image</kbd>
<kbd>image</kbd>

Fitted:

<kbd>image</kbd>
<kbd>image</kbd>

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}

```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the Polaris UI kit

### CODE REVIEW

**Comment by LauraAubin** [comment]
## UX review questions:

- [ ] When tabs are wrapped, the selected tab's underline feels strange.

image

<hr>

### Fitted tabs
- [ ] Keep corners rounded? (both the underline and the tab itself)
- [ ] Keep top padding on container?

image

<hr>

- [ ] Add strong hover state to ellipses? 

image

File: src/components/Tabs/Tabs.scss [hunk 1]
```diff
@@ -3,13 +3,18 @@
...
+$tab-separation-spacing: spacing(tight);
+$tab-item-margin-top: 0.1rem;
 $focus-state-box-shadow-color: rgba(92, 106, 196, 0.8);
 
 .Tabs {
   display: flex;
   flex-wrap: wrap;
   margin: 0;
-  padding: 0;
+  padding: var(--p-border-radius-wide, 0);
```
Comment (LauraAubin) [inline src/components/Tabs/Tabs.scss]: Note:

This should be cleaned up so that we aren't using `border-radius-wide` and instead apply these styles within the `newDesignLanguage` class.

File: src/components/Tabs/Tabs.scss [hunk 2]
```diff
@@ -19,6 +24,7 @@ $focus-state-box-shadow-color: rgba(92, 106, 196, 0.8);
 
   .TabContainer {
     flex: 1 1 100%;
+    margin-left: var(--p-border-radius-wide, 0);
```
Comment (LauraAubin) [inline src/components/Tabs/Tabs.scss]: Same as https://github.com/Shopify/polaris-react/pull/2744/files#r385335917

File: src/components/Tabs/Tabs.scss [hunk 3]
```diff
@@ -244,3 +304,11 @@ $focus-state-box-shadow-color: rgba(92, 106, 196, 0.8);
   visibility: hidden;
   height: 0;
 }
+
+// stylelint-disable selector-max-class
+.newDesignLanguage {
+  padding: $tab-separation-spacing spacing()
+    ($tab-separation-spacing - $border-bottom-height) spacing();
+  margin-left: $tab-separation-spacing;
+  margin-bottom: spacing(extra-tight);
```
Comment (LauraAubin) [inline src/components/Tabs/Tabs.scss]: Note:

Can we de-globalize these?

File: src/components/Tabs/Tabs.tsx [hunk 4]
```diff
@@ -113,9 +118,12 @@ class TabsInner extends React.PureComponent<CombinedProps, State> {
       disclosureActivatorVisible && styles.fillSpace,
     );
 
+    const activatorClassName = classNames(styles.Title, styles.activator);
+
     const disclosureTabClassName = classNames(
       styles.DisclosureTab,
       disclosureActivatorVisible && styles['DisclosureTab-visible'],
+      newDesignLanguage && styles['DisclosureTab-newDesignLanguage'],
```
Comment (LauraAubin) [inline src/components/Tabs/Tabs.tsx]: Note:

Would prefer this to be,

```
newDesignLanguage & styles.newDesignLanguage`
```

and have a `DisclosureTab` class within `newDesignLanguage`.

File: src/components/Tabs/Tabs.scss (hunk 3 above)
Comment (dleroux) [reply src/components/Tabs/Tabs.scss]: What do you mean? Not used the function?

File: src/components/Tabs/Tabs.tsx (hunk 4 above)
Comment (dleroux) [reply src/components/Tabs/Tabs.tsx]: me too :) but we haven't been consistent.

File: src/components/Tabs/Tabs.scss [hunk 5]
```diff
@@ -3,13 +3,18 @@
 $item-min-height: rem(16px);
 $item-min-width: rem(50px);
 $item-vertical-padding: $item-min-height / 2;
+$border-bottom-height: border-width(thicker);
+$tab-separation-spacing: spacing(tight);
+$tab-item-margin-top: 0.1rem;
```
Comment (dleroux) [inline src/components/Tabs/Tabs.scss]: rem(1px) ?

File: src/components/Tabs/Tabs.scss (hunk 1 above)
Comment (dleroux) [reply src/components/Tabs/Tabs.scss]: Yes, I agree, this should just be using `spacing(tight)`. Using this custom property would be a little unpredictable.

File: src/components/Tabs/Tabs.scss [hunk 6]
```diff
@@ -27,8 +33,13 @@ $focus-state-box-shadow-color: rgba(92, 106, 196, 0.8);
 
   .Title {
     width: 100%;
+    margin-left: 0;
```
Comment (dleroux) [inline src/components/Tabs/Tabs.scss]: Why is this needed now?

File: src/components/Tabs/components/Tab/Tab.tsx [hunk 7]
```diff
@@ -108,7 +119,8 @@ export function Tab({
       aria-label={accessibilityLabel}
       onMouseUp={handleMouseUpByBlurring}
     >
-      <span className={styles.Title}>{children}</span>
+      <span className={tabTitleClassNames}>{children}</span>
+      {newDesignLanguage && <div className={selectedUnderlineClassName} />}
```
Comment (dleroux) [inline src/components/Tabs/components/Tab/Tab.tsx]: This should be a `span` because a `div` is a block element and a button is inline-block. Have you tried simply using a `before` selector? It seems like a good candidate for this and reduces the mark-up necessary.

File: src/components/Tabs/Tabs.scss (hunk 1 above)
Comment (danrosenthal) [reply src/components/Tabs/Tabs.scss]: Let's get that change made, and use `spacing(base)`. No need to mis-use a token.

File: src/components/Tabs/components/Tab/Tab.tsx (hunk 7 above)
Comment (danrosenthal) [reply src/components/Tabs/components/Tab/Tab.tsx]: +1 to using a `before` pseudo element for this, as it's presentational in nature

**Comment by dleroux** [comment]
I'm not sure what is causing this, but some reason I can't navigate the tabs using the keyboard in the PR but I am able to on master.

Also, wondering if the `focus-ring` should be coming into play here?

File: src/components/Tabs/Tabs.scss (hunk 3 above)
Comment (LauraAubin) [reply src/components/Tabs/Tabs.scss]: De-globalize in that these styles are put within a class such as `Tab` or `title` (haven't dug into which one specifically)

**Comment by LauraAubin** [comment]
Will re-ping for reviews when this PR is ready, for now I'm going to be pushing up small commits.

File: src/components/Tabs/Tabs.scss (hunk 6 above)
Comment (LauraAubin) [reply src/components/Tabs/Tabs.scss]: No longer needed.

**Comment by dleroux** [comment]
Let me know when this is ready to review again by refreshing my review status.

**Comment by tmlayton** [comment]
Hey @LauraAubin, lmk if you would like to pair through getting the tab measurer sorted. Pending that fix, this is looks great and is good to ship

File: src/components/Tabs/Tabs.scss [hunk 8]
```diff
@@ -244,3 +291,63 @@ $focus-state-box-shadow-color: rgba(92, 106, 196, 0.8);
...
+      .Tab {
+        padding-left: spacing(tight);
+      }
+    }
+    > li:nth-last-child(2) {
+      .Tab {
+        padding-right: spacing(tight);
+      }
+    }
+  }
```
Comment (dleroux) [inline src/components/Tabs/Tabs.scss]: We never know if the activator is visible or not so I'm adding the extra padding to the last Tab, and then removing it from the left of the disclosure.

**Comment by dleroux** [comment]
Removing myself as a reviewer since I completed the PR.

**Comment by tmlayton** [comment]
I made a couple updates—the measuring was still off (assuming the first of and last of type spacing was ignored) so now wrapping tabs with a div which has the left and right spacing as well as the bottom border. This keeps the wrapper spacing outside of the tab measuring logic. Also updated the disclosure icon colors.

**Comment by tmlayton** [comment]
2 visual changes but they were negligible. Appears to be sub pixel shifting based on rendering the border on a wrapping div instead of the ul.
